google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java

```diff
@@ -550,15 +634,67 @@ public void individualWritesErrorIfBulkCommitFails() throws Exception {
     ApiFuture<WriteResult> result2 = bulkWriter.set(doc2, LocalFirestoreHelper.SINGLE_FIELD_MAP);
     bulkWriter.close().get();
 
-    String expected = "java.lang.IllegalStateException: Mock batchWrite failed in test";
     for (ApiFuture<WriteResult> result : Arrays.asList(result1, result2)) {
       try {
         result.get();
       } catch (Exception e) {
-        assertEquals(expected, e.getMessage());
+        assertTrue(
+            e.getMessage()
+                .contains("java.lang.IllegalStateException: Mock batchWrite failed in test"));
         ++opCount;
       }
     }
     assertEquals(2, opCount);
   }
+
+  @Test
+  public void retriesWritesWhenBatchWriteFailsWithRetryableError() throws Exception {
+    FirestoreException retryableError =
+        FirestoreException.serverRejected(
+            Status.fromCode(Status.Code.ABORTED), "Mock batchWrite failed in test");
+
+    ApiFuture<Void> errorFuture = ApiFutures.immediateFailedFuture(retryableError);
+
+    doReturn(errorFuture)
+        .doReturn(errorFuture)
+        .doReturn(errorFuture)
+        .doReturn(successResponse(3))
+        .when(firestoreMock)
+        .sendRequest(
+            batchWriteCapture.capture(),
+            Matchers.<UnaryCallable<BatchWriteRequest, BatchWriteResponse>>any());
+
+    ApiFuture<WriteResult> result = bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.close().get();
+
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(3, 0), result.get().getUpdateTime());
+  }
+
+  @Test
+  public void failsWritesAfterAllRetryAttemptsFail() throws Exception {
+    final int[] retryAttempts = {0};
+    doAnswer(
+            new Answer() {
+              public ApiFuture<Object> answer(InvocationOnMock mock) {
+                retryAttempts[0]++;
+                return ApiFutures.immediateFailedFuture(
+                    FirestoreException.serverRejected(
+                        Status.fromCode(Status.Code.ABORTED), "Mock batchWrite failed in test"));
+              }
+            })
+        .when(firestoreMock)
+        .sendRequest(
+            batchWriteCapture.capture(),
+            Matchers.<UnaryCallable<BatchWriteRequest, BatchWriteResponse>>any());
+    ApiFuture<WriteResult> result = bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.close().get();
+
+    try {
+      result.get();
+      Assert.fail("Expected set() operation to fail");
+    } catch (Exception e) {
+      assertTrue(e.getMessage().contains("Mock batchWrite failed in test"));
+      assertEquals(retryAttempts[0], BulkWriter.MAX_RETRY_ATTEMPTS + 1);
+    }
+  }
```

Comment:
Can we add a test that:

- Adds two writes
- Calls flush
- Fails one write with retryable error, let's the other succeed
- Retries write which succeeds
- Test verifies that flush() only finishes now

I am not 100% convinced that this works right now.

Reply:
Added test.